Automattic/woocommerce-payments
# Better handling of failed payment_intent webhooks.

Fixes #3646 

#### Changes proposed in this Pull Request

* Update the `process_webhook_payment_intent_failed` method.
  * Check to make sure the payment method type in the intent is a type we process failed intents for.
  * Check to make sure the payment method in the intent matches what is in the order. This is due to methods could leave the order as Pending Payment, which allows the customer to use a different method type to pay the order. We wouldn't want to allow the order to fail if it was already successful. Checking the order status alone doesn't help due to there are extensions to add order statuses. 
* Add `WC_Payments_Utils::compose_transaction_url` method, which was just copied from the main `WC_Payment_Gateway_WCPay` class.
* Add `WC_Payments_Utils::mark_payment_failed` method to mark a payment (order) as failed.
  * Creates note stating that the payment has failed, then adds any additional message passed to the method onto the note. 
  * Updates the `_intention_status` meta of the order to `failed`.
  * Updates the order status to Failed.
* Updated tests.

#### Testing instructions

* Tests should pass.

I found the easiest way to manually test is to use Postman, unless you have another method for sending in requests to the REST API.

* Use filter below to get around authentication, be sure to remove it afterwards.
* Set to POST to `https://YOURSITE/?rest_route=/wc/v3/payments/webhook`.
* Use the json below as the body of the post, updating where needed.
  * It's best to use an existing order, and get the payment intent and payment method ids out of it, then add those to the json data.
  * Set the order to Pending.
* After posting:
  * The order should be moved to Failed status.
  * There should be a note in the order that the payment failed with a link to the transactions, and it should state the account was not found.
  * There should be a note stating the order status changed from Pending to Failed.
  * Example screenshot is below.

filter:
```
add_filter( 'user_has_cap', function( $caps ) {
	$caps['manage_woocommerce'] = true;
	return $caps;
});
```

json:
```
{
    "type": "payment_intent.payment_failed",
    "data": {
        "object": {
            "id": "ADD_INTENT_ID_HERE",
            "object": "payment_intent",
            "charges": {
                "data": [
                    {
                        "failure_code": "no_account",
                        "payment_method": "ADD_PAYMENT_METHOD_ID_HERE",
                        "payment_method_details": {
                            "type": "us_bank_account"
                        }
                    }
                ]
            }
        }
    }
}
```

![Screenshot on 2022-01-17 at 14-03-39](https://user-images.githubusercontent.com/4634416/149819343-2d70033d-bb44-4337-89f0-6246f6c78475.png)


-------------------


**Post merge**
